[MDEV-16735] Ensure mysql_upgrade works when changing alter_algorithm

christian7007 opened the pull request:
## Description

MDEV-16735 describes how mysql_upgrade fails when alter_algorithm
is set to a value different than DEFAULT/COPY. It was marked as
fixed by 0ee08683, but there was a typo in it. This commit
fix the typo and add MTRs to ensure the fix is working properly.

```
MariaDB [(none)]> set alter_algorithm=DEFAULT;
Query OK, 0 rows affected (0.000 sec)

MariaDB [(none)]> show variables like 'alter_algorithm';
+-----------------+---------+
| Variable_name   | Value   |
+-----------------+---------+
| alter_algorithm | INPLACE |
+-----------------+---------+
1 row in set (0.001 sec)

MariaDB [(none)]> set alter_algorithm='DEFAULT';
Query OK, 0 rows affected (0.000 sec)

MariaDB [(none)]> show variables like 'alter_algorithm';
+-----------------+---------+
| Variable_name   | Value   |
+-----------------+---------+
| alter_algorithm | DEFAULT |
+-----------------+---------+
1 row in set (0.001 sec)
```

**NOTE:** I will work on a follow up PR with the changes for making `set alter_algorithm=DEFAULT;` query fail with the corresponding error message.

## How can this PR be tested?
The corresponding MTR test has been added, so just running the MTR tests as below should be enough (any other test should work as before):

```
./mtr main.mysql_upgrade
```

## Basing the PR against the correct MariaDB version
- [x] *This is a bug fix and the PR is based against the earliest branch in which the bug can be reproduced*

## Copyright
All new code of the whole pull request, including one or several files that are either new files or modified ones, are contributed under the BSD-new license. I am contributing on behalf of my employer Amazon Web Services, Inc.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/MariaDB/server?pullRequest=2345) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/MariaDB/server?pullRequest=2345) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/MariaDB/server?pullRequest=2345) it.</sub>

grooverdan commented on mysql-test/main/mysql_upgrade.test:
```diff
@@ -472,4 +472,15 @@ drop table mysql.global_priv;
 --remove_file $MYSQLD_DATADIR/mysql_upgrade_info
 
+--echo #
+--echo # [MDEV-16735] Upgrades only work if 'alter_algorithm' is DEFAULT
```
remove `[]` for consistency with other tests.

grooverdan commented on mysql-test/main/mysql_upgrade.test:
```diff
@@ -472,4 +472,15 @@ drop table mysql.global_priv;
+SET GLOBAL alter_algorithm='INPLACE';
+SHOW GLOBAL VARIABLES LIKE 'alter_algorithm';
+--exec $MYSQL_UPGRADE --force 2>&1
+SET GLOBAL alter_algorithm='DEFAULT';
```
Actually here you want DEFAULT unquoted. In case, for what will be horribly readable reasons, the `alter_algorithm` is changed.

grooverdan reviewed (changes requested):
Note unquoted DEFAULT wasn't so much a typo as a wrong assumption that no-one would change the alter_algorithm globally. E.g. globally set to `INPLACE` is the behaviour:

```
MariaDB [(none)]> show variables like 'alter_algorithm';
+-----------------+---------+
| Variable_name   | Value   |
+-----------------+---------+
| alter_algorithm | INPLACE |
+-----------------+---------+
1 row in set (0.001 sec)

MariaDB [(none)]> set alter_algorithm='INSTANT'
    -> ;
Query OK, 0 rows affected (0.000 sec)

MariaDB [(none)]> show variables like 'alter_algorithm';
+-----------------+---------+
| Variable_name   | Value   |
+-----------------+---------+
| alter_algorithm | INSTANT |
+-----------------+---------+
1 row in set (0.001 sec)

MariaDB [(none)]> set alter_algorithm=DEFAULT;
Query OK, 0 rows affected (0.000 sec)

MariaDB [(none)]> show variables like 'alter_algorithm';
+-----------------+---------+
| Variable_name   | Value   |
+-----------------+---------+
| alter_algorithm | INPLACE |
+-----------------+---------+
1 row in set (0.001 sec)

MariaDB [(none)]> set alter_algorithm='DEFAULT';
Query OK, 0 rows affected (0.000 sec)

MariaDB [(none)]> show variables like 'alter_algorithm';
+-----------------+---------+
| Variable_name   | Value   |
+-----------------+---------+
| alter_algorithm | DEFAULT |
+-----------------+---------+
1 row in set (0.001 sec)
```

Unquoted default is per https://mariadb.com/kb/en/set/#default.

Its still the right fix in `mysql_system_tables_fix.sql`, so well done.

As 0ee086838d4b4497599d4e0822343269b682f913 was in 10.3, this should be also.

Thanks for the contribution.

christian7007 commented:
Hi @grooverdan thank you for your feedback and for the reference to the `DEFAULT` value behaviour. I've updated the PR with the changes you asked, changed the target branch to 10.3 and added a small fix to the MTR test to removes the `mysql_upgrade_info` file, it seems I missed the check which complains about it while preparing my previous commit.

grooverdan commented:
Thanks for the force push. I was wondering why I didn't see the option for it.

Thanks for the fix too.

christian7007 commented:
No problem, thank you for the quick review!
